tests/misc/tlifted_lambda_issue.nim

```diff
@@ -0,0 +1,28 @@
+discard """
+  knownIssue: '''
+    variables defined in statement list that are later lifted into lambda
+    expressions are first semantically analysed as normal module-level code,
+    causing them to stay as globals, which is invalid and confuses the
+    compiler backend
+  '''
+"""
+
+# https://github.com/nim-lang/nim/issues/7104
+
+var output: seq[int]
+
+proc sp(cb: proc())=
+  cb()
+
+sp:
```

Comment:
The elif branch in `sigmatch.paramTypesMatchAux` that uses `matchesVoidProc` is where lifting a statement list into a lambda expression happens.

Reply:
Crap, I lost my reply. :( Shortened version, of what I recall.

I think the issue is `prepareOperand` needs to store more than one analysis per argument, because semanticizing for a formal parameter that's lifted to a procedural isn't the same as say `typed`. So we should do the conversion earlier and store the conversion for the arg in its list of analysed possibilities. Then depending upon which one we choose to dispatch, we pick that analysis. Something like `seq[array[AnalysisKind, PNode]]`, where `seq.len == args.len` and `AnalysisKind` covers, concrete, iterator, meta, typed, procedural, untyped, etc.

The thing is shadow scopes juggling becomes bonkers, we might need to create them as nested based on unique sequence of analyses. So we can clearly discard things.